fix: Move temporary files outside of the cache directory

kitadai31 opened the pull request:
fix for #2120

The "ephemeral" directory is moved from cacheDir to filesDir.
The input.apk (exists if selected from storage) is moved to the ephemeral directory.

I'm not sure if the changes in `Session.kt` are right.
This is intended to delete input.apk as soon as possible when it is no longer needed.
If the apk is a downloaded apk or selected from installed apps, nothing happens on that line.

oSumAtrIX commented:
Does this PR also delete the temporal files after they are not necessary anymore?

kitadai31 commented:
Yes
After patching is finished, only the output patched apk will be in the temporary folder
This apk will also be deleted when I back to the dashboard

oSumAtrIX commented:
Which line of code deletes the temp files?

kitadai31 commented:
`File(cacheDir, "input.apk").delete()` removes a input apk which is selected from storage

Other temporary files are already deleted after patching since before this PR

Axelen123 commented on app/src/main/java/app/revanced/manager/data/platform/Filesystem.kt:
```diff
@@ -16,7 +16,7 @@ class Filesystem(private val app: Application) {
      * Do not store paths to this directory in a parcel.
      */
-    val tempDir = app.cacheDir.resolve("ephemeral").apply {
+    val tempDir = app.filesDir.resolve("ephemeral").apply {
```
Use getDir instead

kitadai31 replied:
Since getDir creates the directory when it is called, I thought it was not suitable because I would have to create the directory myself.
I noticed that getDir and mkdirs() have different permissions on the folders created.
Should I still use getDir?

Axelen123 replied:
Yes, because that is what the android docs recommends.

Axelen123 commented on app/src/main/java/app/revanced/manager/patcher/Session.kt:
```diff
@@ -129,6 +129,7 @@ class Session(
 
     override fun close() {
         tempDir.deleteRecursively()
+        File(cacheDir, "input.apk").delete()
```
PatcherViewModel already deletes it.

Axelen123 commented on app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:
```diff
@@ -17,15 +18,19 @@ import kotlinx.coroutines.Dispatchers
+) : ViewModel(), KoinComponent {
+    private val fs: Filesystem by inject()
+
+    private val inputFile = File(fs.tempDir, "input.apk").also {
```
tempDir is only valid for the duration of the app process, but this needs to live longer than that. Use getDir instead.

kitadai31 commented:
Reflected the review, thank you.

Also, I moved the deleting selected apk code to PatcherWorker.

Currently, PatcherViewModel deletes the selected apk on `onCleared()` method.
But this may cause the selected apk (input.apk) to remain in internal storage because onCleared() won't be called if the users exit the task from recent apps.
If user exits the task after patching has finished without returning to the dashboard by accident, the input.apk remains in the app data.
Therefore, I think the input.apk should be deleted as soon as possible when patching is ended.
This is the reason I moved the code.

Is this OK?

Axelen123 commented on app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:
```diff
@@ -25,7 +26,7 @@ class AppSelectorViewModel(
     private val patchBundleRepository: PatchBundleRepository
 ) : ViewModel() {
-    private val inputFile = File(app.cacheDir, "input.apk").also {
+    private val inputFile = File(app.getDir("ephemeral", Context.MODE_PRIVATE), "input.apk").also {
```
"ephemeral" is deleted by `Filesystem` every time the app process is restarted. Use another directory.

kitadai31 replied:
input.apk should be deleted every time the app restarted, shouldn't it?
Is input.apk needed after restarted?

Axelen123 replied:
We cannot delete the input APK every time the app process is restarted because of [system-initiated process death](https://galex.dev/posts/process-death-is-the-rule-not-the-exception/#users-experience-system-initiated-process-death-a-lot). The process is killed when the app is in the background, but when the user navigates back a new process is launched and state is restored. System-initiated process death would definitely cause problems if the app is killed while the user is in a screen between `AppSelectorScreen` and `PatcherScreen`.

kitadai31 replied:
OK understood.
I now get what you mean by "process".

Since this is a single file, I think the destination will be `filesDir`. Is that correct?
( `File(app.filesDir, "input.apk")` )

Axelen123 replied:
Use something like `getDir("app_selector", Context.MODE_PRIVATE)`. I remember having problems when trying to use `filesDir` directly.

kitadai31 replied:
Isn't it [`dataDir`](https://developer.android.com/reference/android/content/Context#getDataDir()) ?

> Apps should not use this path directly; they should instead use getFilesDir(), getCacheDir(), getDir(java.lang.String, int), or other storage APIs on this class. 

`filesDir` is a standard location to place a file
https://developer.android.com/training/data-storage/app-specific#internal-access-store-files
I replaced it with `File(app.filesDir, "input.apk")` and there were no problems.

Axelen123 replied:
In that case `filesDir` is fine.

Axelen123 commented on app/src/main/java/app/revanced/manager/patcher/worker/PatcherWorker.kt:
```diff
@@ -193,6 +193,9 @@ class PatcherWorker(
             Result.failure()
         } finally {
             patchedApk.delete()
+            if (args.input is SelectedApp.Local && args.input.temporary) {
+                args.input.file.delete()
+            }
```
What is the reason for this change?

kitadai31 replied:
Please read the message above 🙇‍♂️

Axelen123 replied:
My bad, makes sense.
